userbenchmark/ddp_experiments/__init__.py

```diff
@@ -0,0 +1,283 @@
+    parser.add_argument(
+        "--index_file",
+        type=str,
+        default=f"ddp_experiments_{datetime.now().strftime('%Y%m%d-%H%M%S')}.csv",
+        help="training paradigm, by default using DDP"
+    )
+
+
+    try:
+        if args:
+            return parser.parse_args(args)
+        else:
+            return parser.parse_args()
+    except:
+        parser.print_help()
+        sys.exit(0)
+
+
+def get_init_file(args):
+    # Init file must not exist, but it's parent dir must exist.
+    os.makedirs(args.job_dir, exist_ok=True)
+    init_file = Path(args.job_dir) / f"{uuid.uuid4().hex}_init"
+    print(init_file)
+    if init_file.exists():
+        os.remove(str(init_file))
+    return init_file
+
+
+class TrainerWrapper(object):
+    def __init__(self, args, model_args):
+        self.args = args
+        self.args.output_dir = args.job_dir
+        
+        # extra args just passed to the Trainer class ctor
+        self.model_args=model_args
+
+    def __call__(self):
+        self._setup_gpu_args()
+
+        pos = self.args.model.rfind(".")
+        module = importlib.import_module(self.args.model[:pos])
+        model_class = getattr(module, self.args.model[(pos+1):])
+
+        pos = self.args.trainer.rfind(".")
+        module = importlib.import_module(self.args.trainer[:pos])
+        trainer_class = getattr(module, self.args.trainer[(pos+1):])
+
+        return trainer_class(self.args, model_class, model_args=self.model_args).measure()
+
+    def checkpoint(self):
+        self.args.dist_url = get_init_file(self.args).as_uri()
+        checkpoint_file = os.path.join(self.args.output_dir, "checkpoint.pth")
+        if os.path.exists(checkpoint_file):
+            self.args.resume = checkpoint_file
+        print("Requeuing ", self.args)
+        empty_trainer = type(self)(self.args)
+        return submitit.helpers.DelayedSubmission(empty_trainer)
+
+    def _setup_gpu_args(self):
+        job_env = submitit.JobEnvironment()
+        self.args.output_dir = Path(str(self.args.output_dir).replace("%j", str(job_env.job_id)))
+        self.args.gpu = job_env.local_rank
+        self.args.rank = job_env.global_rank
+        self.args.world_size = job_env.num_tasks
+        print(f"Process group: {job_env.num_tasks} tasks, rank: {job_env.global_rank}")
+
+        os.environ["LOCAL_RANK"] = str(job_env.local_rank)
+        os.environ["RANK"] = str(job_env.global_rank)
+        os.environ["WORLD_SIZE"] = str(job_env.num_tasks)
+        os.environ["GPUS_PER_NODE"] = str(job_env.num_tasks//job_env.num_nodes)
+        # os.environ["NCCL_IB_DISABLE"] = str(1)
+        os.environ["NCCL_DEBUG"] = 'INFO'
+        os.environ["NCCL_DEBUG_SUBSYS"] = 'INIT,ENV,NET'
+        os.environ['NCCL_SOCKET_IFNAME'] = 'ens'
+        # os.environ["NCCL_ALGO"] = 'ring'
+        os.environ["FI_PROVIDER"] = 'efa'
+        os.environ["FI_EFA_USE_DEVICE_RDMA"]= str(1)
+        os.environ["NET_TYPE"] = 'efa'
+
+
+def main():
+    args = parse_args()
```

Comment:
Can we have an example command to run this userbenchmark and an output json text? Would like to know how it looks like.

Reply:
Added a README.md (modeled after the distributed userbenchmark) - does that work?

Reply:
Yes, adding `README.md` sounds good to me.